### PR TITLE
ci: run extension tests

### LINKED ISSUE #11: Build up CI test environment
We should have defined GitHub workflows for checking compliance of commits in PR and make tests before merging, like in [bee-js](https://github.com/ethersphere/bee-js), but for only (chrome) browser environment, where the tests can run in GUI mode.

### PR DESCRIPTION
**Pull request by vojtechsimetka**
Resolves #11

### CODE REVIEW

**Comment by nugaon** [comment]
It tries to call `npm run compile:types` which does not exist anymore.

**Review by nugaon — COMMENTED** [review]
The image `mujo-code/puppeteer-headful` that you are using requires to change the puppeteer launch configuration:

> Then you will need to change the way you launch Puppeteer. We export out a nifty ENV variable PUPPETEER_EXEC_PATH that you set at your executablePath. This should be undefined locally so it should function perfectly fine locally and on the action.

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  push:
+    branches:
+      - 'master'
+  pull_request:
+    branches:
+      - '**'
+
+env:
+  BEE_VERSION: '1.1.0'
+  BLOCKCHAIN_VERSION: '1.2.0-for-1.1.0'
```
[the remark this answers is not in the document]
Comment (vojtechsimetka) [reply .github/workflows/tests.yaml]: You had outdated code loaded for some reason :)

File: test/config/setup.ts [hunk 2]
```diff
@@ -9,8 +9,9 @@ const EXTENSION_PATH = path.join(__dirname, '..', '..', '/dist')
 module.exports = async () => {
   console.log('Setup Puppeteer')
   const browser = await puppeteer.launch({
+    executablePath: process.env.PUPPETEER_EXEC_PATH, // set by docker container
     headless: false, // extensions only supported in full chrome.
-    args: [`--disable-extensions-except=${EXTENSION_PATH}`, `--load-extension=${EXTENSION_PATH}`],
+    args: [`--disable-extensions-except=${EXTENSION_PATH}`, `--load-extension=${EXTENSION_PATH}`, '--no-sandbox'],
```
Comment (nugaon) [inline test/config/setup.ts:14]: why did you add `--no-sandbox`?

Comment (vojtechsimetka) [reply test/config/setup.ts:14]: Pretty much just to see if it is also required and can solve our issue. Removing it now.

**Comment by vojtechsimetka** [comment]
So I am abandoning this issue for now. The commit 4e2decd starts the puppeteer correctly but would need further debugging. I've encountered these errors in past locally as well but no idea what actually solved it. For more info see the run https://github.com/ethersphere/swarm-extension/runs/3688641110

Here is a complete log:
```
Run mujo-code/puppeteer-headful@master
/usr/bin/docker run --name e1cc516104dabbf40b42929a4e0bafc7b1bb5c_8ff6f5 --label e1cc51 --workdir /github/workspace --rm -e BEE_VERSION -e BLOCKCHAIN_VERSION -e BEE_ENV_PREFIX -e BEE_IMAGE_PREFIX -e COMMIT_VERSION_TAG -e WORKERS -e CI -e INPUT_ARGS -e HOME -e GITHUB_JOB -e GITHUB_REF -e GITHUB_SHA -e GITHUB_REPOSITORY -e GITHUB_REPOSITORY_OWNER -e GITHUB_RUN_ID -e GITHUB_RUN_NUMBER -e GITHUB_RETENTION_DAYS -e GITHUB_RUN_ATTEMPT -e GITHUB_ACTOR -e GITHUB_WORKFLOW -e GITHUB_HEAD_REF -e GITHUB_BASE_REF -e GITHUB_EVENT_NAME -e GITHUB_SERVER_URL -e GITHUB_API_URL -e GITHUB_GRAPHQL_URL -e GITHUB_WORKSPACE -e GITHUB_ACTION -e GITHUB_EVENT_PATH -e GITHUB_ACTION_REPOSITORY -e GITHUB_ACTION_REF -e GITHUB_PATH -e GITHUB_ENV -e RUNNER_OS -e RUNNER_NAME -e RUNNER_TOOL_CACHE -e RUNNER_TEMP -e RUNNER_WORKSPACE -e ACTIONS_RUNTIME_URL -e ACTIONS_RUNTIME_TOKEN -e ACTIONS_CACHE_URL -e GITHUB_ACTIONS=true -v "/var/run/docker.sock":"/var/run/docker.sock" -v "/home/runner/work/_temp/_github_home":"/github/home" -v "/home/runner/work/_temp/_github_workflow":"/github/workflow" -v "/home/runner/work/_temp/_runner_file_commands":"/github/file_commands" -v "/home/runner/work/swarm-extension/swarm-extension":"/github/workspace" e1cc51:6104dabbf40b42929a4e0bafc7b1bb5c npm run test:ci
Running 'npm run test:ci'!

> @ethersphere/swarm-extension@0.1.0 test:ci /github/workspace
> BEE_API_URL=http://172.18.0.1:1633 BEE_DEBUG_API_URL=http://172.18.0.1:1635 jest --config=jest.config.ts

Buying stamp on the Bee node http://172.18.0.1:1635...
Waiting 11 secs for batch ID settlement...
Bee postage stamp: bb2c02e00decea39f2d3378636864e4d294a325a178c21d1f76a5849515711bf 
 You can set it to BEE_STAMP sys env variable to don't wait again for buying one
Setup Puppeteer
  console.error
    Unhandled error

      at process.uncaught (../node_modules/jest-jasmine2/build/jasmine/Env.js:248:21)
      at process.emit (../node_modules/source-map-support/source-map-support.js:495:21)

  console.error
    TypeError: n.default.promises.opendir is not a function
        at buildCollectionRelative (/github/workspace/node_modules/@ethersphere/bee-js/dist/webpack:/BeeJs/src/utils/collection.ts:43:37)
        at makeCollectionFromFS (/github/workspace/node_modules/@ethersphere/bee-js/dist/webpack:/BeeJs/src/utils/collection.ts:37:10)
        at t.Bee.uploadFilesFromDirectory (/github/workspace/node_modules/@ethersphere/bee-js/dist/webpack:/BeeJs/src/bee.ts:284:24)
        at uploadFilesFromDirectory (/github/workspace/test/bzz-protocol.spec.ts:86:18)
        at Object.<anonymous> (/github/workspace/test/bzz-protocol.spec.ts:88:28)
        at /github/workspace/node_modules/jest-jasmine2/build/queueRunner.js:45:12
        at new Promise (<anonymous>)
        at mapper (/github/workspace/node_modules/jest-jasmine2/build/queueRunner.js:28:19)
        at /github/workspace/node_modules/jest-jasmine2/build/queueRunner.js:75:41
        at processTicksAndRejections (internal/process/task_queues.js:93:5)

      at process.uncaught (../node_modules/jest-jasmine2/build/jasmine/Env.js:249:21)
      at process.emit (../node_modules/source-map-support/source-map-support.js:495:21)

FAIL test/bzz-protocol.spec.ts (75.123 s)
  BZZ protocol
    ✕ checks inner iframe reference load (1005 ms)
    ✕ checks iframe has been loaded with web+bzz reference (1003 ms)
    ✕ Fetch Real Bee API URL (1002 ms)
    ✕ Allow Global Postage Stamp ID (1066 ms)
    ✕ Upload file through Fake URL (1002 ms)
    ✕ Fetch image via Fake URL (1001 ms)
    ✕ check {cid}.bzz.link image resource loaded (1004 ms)
    ✕ click on web+bzz link reference (1003 ms)
    ✕ type in bzz://{content-id} address into the address bar with default search engine Google (1586 ms)
    ✕ type in https://{cid}.bzz.link address into the address bar (1003 ms)
    ✕ Change Bee API URL (1061 ms)
    ✕ Check traditional and swarm localStorage usage (31009 ms)
    ✕ checks window object and content document is not available of iframe because of cross-origin (1564 ms)

  ● BZZ protocol › checks inner iframe reference load

    Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.

      at mapper (../node_modules/jest-jasmine2/build/queueRunner.js:27:45)

  ● BZZ protocol › checks inner iframe reference load

    TypeError: Cannot read property '$' of undefined

      106 |
      107 |   test('checks inner iframe reference load', async done => {
    > 108 |     const ref = await page.$('#localhost-inner-ref')
          |                            ^
      109 |
      110 |     checkJinnIframePage(ref)
      111 |     done()

      at Object.<anonymous> (bzz-protocol.spec.ts:108:28)

  ● BZZ protocol › checks iframe has been loaded with web+bzz reference

    Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.

      at mapper (../node_modules/jest-jasmine2/build/queueRunner.js:27:45)

  ● BZZ protocol › checks iframe has been loaded with web+bzz reference

    TypeError: Cannot read property '$' of undefined

      113 |
      114 |   test('checks iframe has been loaded with web+bzz reference', async () => {
    > 115 |     const ref = await page.$('#bzz-iframe')
          |                            ^
      116 |
      117 |     checkJinnIframePage(ref)
      118 |   })

      at Object.<anonymous> (bzz-protocol.spec.ts:115:28)

  ● BZZ protocol › Fetch Real Bee API URL

    Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.

      at mapper (../node_modules/jest-jasmine2/build/queueRunner.js:27:45)

  ● BZZ protocol › Fetch Real Bee API URL

    TypeError: Cannot read property 'click' of undefined

      119 |
      120 |   test('Fetch Real Bee API URL', async done => {
    > 121 |     await page.click('#button-fetch-real-bee-api-url')
          |                ^
      122 |     const placeHolderSelector = '#bee-api-url-placeholder'
      123 |     await page.waitForSelector(placeHolderSelector)
      124 |     const value = await page.$eval(placeHolderSelector, e => e.innerHTML)

      at Object.<anonymous> (bzz-protocol.spec.ts:121:16)

  ● BZZ protocol › Allow Global Postage Stamp ID

    Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.

      at mapper (../node_modules/jest-jasmine2/build/queueRunner.js:27:45)

  ● BZZ protocol › Allow Global Postage Stamp ID

    net::ERR_BLOCKED_BY_CLIENT at chrome-extension://undefined/popup-page/index.html

      at navigate (../node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:115:23)

  ● BZZ protocol › Upload file through Fake URL

    Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.

      at mapper (../node_modules/jest-jasmine2/build/queueRunner.js:27:45)

  ● BZZ protocol › Upload file through Fake URL

    TypeError: Cannot read property 'click' of undefined

      151 |
      152 |   test('Upload file through Fake URL', async done => {
    > 153 |     await page.click('#button-upload-fake-url-file')
          |                ^
      154 |     const placeHolderSelector = '#fake-bzz-url-content-1 > a:first-child'
      155 |     await page.waitForSelector(placeHolderSelector)
      156 |     await page.click(placeHolderSelector)

      at Object.<anonymous> (bzz-protocol.spec.ts:153:16)

  ● BZZ protocol › Fetch image via Fake URL

    Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.

      at mapper (../node_modules/jest-jasmine2/build/queueRunner.js:27:45)

  ● BZZ protocol › Fetch image via Fake URL

    TypeError: Cannot read property 'click' of undefined

      163 |
      164 |   test('Fetch image via Fake URL', async done => {
    > 165 |     await page.click('#button-fetch-jinn-page')
          |                ^
      166 |     const placeHolderSelector = '#fake-url-fetch-jinn > img:first-child'
      167 |     await page.waitForSelector(placeHolderSelector)
      168 |     const completedImageLoad = await page.evaluate(async placeHolderSelector => {

      at Object.<anonymous> (bzz-protocol.spec.ts:165:16)

  ● BZZ protocol › check {cid}.bzz.link image resource loaded

    Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.

      at mapper (../node_modules/jest-jasmine2/build/queueRunner.js:27:45)

  ● BZZ protocol › check {cid}.bzz.link image resource loaded

    TypeError: Cannot read property 'evaluate' of undefined

      180 |
      181 |   test('check {cid}.bzz.link image resource loaded', async () => {
    > 182 |     const imageLoaded = await page.evaluate(() => {
          |                                    ^
      183 |       return (document.getElementById('bzz-image-2') as HTMLImageElement).complete
      184 |     })
      185 |

      at Object.<anonymous> (bzz-protocol.spec.ts:182:36)

  ● BZZ protocol › click on web+bzz link reference

    Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.

      at mapper (../node_modules/jest-jasmine2/build/queueRunner.js:27:45)

  ● BZZ protocol › click on web+bzz link reference

    TypeError: Cannot read property 'click' of undefined

      189 |   test('click on web+bzz link reference', async () => {
      190 |     // perform navigation
    > 191 |     await page.click('#bzz-ext-ref')
          |                ^
      192 |     const currentPage = await getLastBzzPage()
      193 |     const jinnPage = await currentPage.$('#jinn-page-title')
      194 |

      at Object.<anonymous> (bzz-protocol.spec.ts:191:16)

  ● BZZ protocol › type in bzz://{content-id} address into the address bar with default search engine Google

    Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.

      at mapper (../node_modules/jest-jasmine2/build/queueRunner.js:27:45)

  ● BZZ protocol › type in bzz://{content-id} address into the address bar with default search engine Google

    expect(received).toBe(expected) // Object.is equality

    Expected: "First Direct BZZ address"
    Received: "localhost"

      202 |     const bzzPageTitleElement = await currentPage.$('#first-bzz-page-title')
      203 |
    > 204 |     expect(title).toBe('First Direct BZZ address')
          |                   ^
      205 |     expect(bzzPageTitleElement).toBeTruthy()
      206 |   })
      207 |

      at Object.<anonymous> (bzz-protocol.spec.ts:204:19)

  ● BZZ protocol › type in https://{cid}.bzz.link address into the address bar

    Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.

      at mapper (../node_modules/jest-jasmine2/build/queueRunner.js:27:45)

  ● BZZ protocol › type in https://{cid}.bzz.link address into the address bar

    TypeError: Cannot read property 'indexOf' of undefined

      64 |
      65 | export function bzzResourceToSubdomain(bzzReference: string): string | null {
    > 66 |   const lastSlash = bzzReference.indexOf('/')
         |                                  ^
      67 |
      68 |   if (lastSlash) bzzReference.slice(0, lastSlash)
      69 |

      at bzzResourceToSubdomain (../src/utils/bzz-link.ts:66:34)
      at Object.<anonymous> (bzz-protocol.spec.ts:209:17)

  ● BZZ protocol › Change Bee API URL

    Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.

      at mapper (../node_modules/jest-jasmine2/build/queueRunner.js:27:45)

  ● BZZ protocol › Change Bee API URL

    net::ERR_BLOCKED_BY_CLIENT at chrome-extension://undefined/popup-page/index.html

      at navigate (../node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:115:23)

  ● BZZ protocol › Check traditional and swarm localStorage usage

    Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.

      at mapper (../node_modules/jest-jasmine2/build/queueRunner.js:27:45)

  ● BZZ protocol › Check traditional and swarm localStorage usage

    TimeoutError: waiting for selector `form[id="form-bee-api-url-change"] input[type="text"]` failed: timeout 30000ms exceeded

      44 |
      45 | export async function getElementBySelector(selector: string, page: Page): Promise<ElementHandle<Element>> {
    > 46 |   await page.waitForSelector(selector)
         |              ^
      47 |   const element = await page.$(selector)
      48 |
      49 |   if (!element) throw new Error(`Element with selector ${selector} has been not found`)

      at new WaitTask (../node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:505:34)
      at DOMWorld.waitForSelectorInPage (../node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:416:26)
      at Object.internalHandler.waitFor (../node_modules/puppeteer/lib/cjs/puppeteer/common/QueryHandler.js:31:77)
      at DOMWorld.waitForSelector (../node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:312:29)
      at Frame.waitForSelector (../node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:842:51)
      at Page.waitForSelector (../node_modules/puppeteer/lib/cjs/puppeteer/common/Page.js:1285:33)
      at getElementBySelector (utils.ts:46:14)
      at changeBeeApiUrl (bzz-protocol.spec.ts:46:27)
      at Object.<anonymous> (bzz-protocol.spec.ts:243:36)

  ● BZZ protocol › checks window object and content document is not available of iframe because of cross-origin

    Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.

      at mapper (../node_modules/jest-jasmine2/build/queueRunner.js:27:45)

  ● BZZ protocol › checks window object and content document is not available of iframe because of cross-origin

    Evaluation failed: Error: there is no "localstorage-iframe" element

      at ../__puppeteer_evaluation_script__:5:15
      at ExecutionContext._evaluateInternal (../node_modules/puppeteer/lib/cjs/puppeteer/common/ExecutionContext.js:217:19)
      at ExecutionContext.evaluate (../node_modules/puppeteer/lib/cjs/puppeteer/common/ExecutionContext.js:106:16)

Test Suites: 1 failed, 1 total
Tests:       13 failed, 13 total
Snapshots:   0 total
Time:        75.206 s
Ran all test suites.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ethersphere/swarm-extension@0.1.0 test:ci: `BEE_API_URL=http://172.18.0.1:1633 BEE_DEBUG_API_URL=http://172.18.0.1:1635 jest --config=jest.config.ts`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @ethersphere/swarm-extension@0.1.0 test:ci script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /github/home/.npm/_logs/2021-09-23T14_18_33_618Z-debug.log
Failure running 'npm run test:ci', exited with 1
```

**Review by Cafe137 — APPROVED** [review]
